app/src/main/java/org/oppia/app/player/state/StateFragmentPresenter.kt

```diff
@@ -139,4 +216,166 @@ class StateFragmentPresenter @Inject constructor(
+    return Transformations.map(answerOutcome, ::processAnswerOutcome)
+  }
+
+  /** Helper for subscribeToAnswerOutcome. */
+  private fun processAnswerOutcome(ephemeralStateResult: AsyncResult<AnswerOutcome>): AnswerOutcome {
+    if (ephemeralStateResult.isFailure()) {
+      logger.e("StateFragment", "Failed to retrieve answer outcome", ephemeralStateResult.getErrorOrNull()!!)
+    }
+    return ephemeralStateResult.getOrDefault(AnswerOutcome.getDefaultInstance())
+  }
+
+  private fun endExploration() {
+    explorationDataController.stopPlayingExploration()
+    (activity as ExplorationActivity).finish()
+  }
+
+  override fun onInteractionButtonClicked() {
+    hideKeyboard()
+    // TODO(#163): Remove these dummy answers and fetch answers from different interaction views.
+    // NB: This sample data will work only with TEST_EXPLORATION_ID_5
+    // 0 -> What Language
+    // 2 -> Welcome!
+    // XX -> What Language
+    val stateWelcomeAnswer = 0
+    // finnish -> Numeric input
+    // suomi -> Numeric input
+    // XX -> What Language
+    val stateWhatLanguageAnswer: String = "finnish"
+    // 121 -> Things You can do
+    // < 121 -> Estimate 100
+    // > 121 -> Numeric Input
+    // XX -> Numeric Input
+    val stateNumericInputAnswer = 121
+
+    if (!hasGeneralContinueButton) {
+      val interactionObject: InteractionObject = getDummyInteractionObject()
+      when (currentEphemeralState.get()!!.state.interaction.id) {
+        END_EXPLORATION -> endExploration()
+        CONTINUE -> subscribeToAnswerOutcome(explorationProgressController.submitAnswer(createContinueButtonAnswer()))
+        MULTIPLE_CHOICE_INPUT -> subscribeToAnswerOutcome(
+          explorationProgressController.submitAnswer(
+            InteractionObject.newBuilder().setNonNegativeInt(
+              stateWelcomeAnswer
+            ).build()
+          )
+        )
+        FRACTION_INPUT,
+        ITEM_SELECT_INPUT,
+        NUMERIC_INPUT,
+        NUMERIC_WITH_UNITS,
+        TEXT_INPUT -> subscribeToAnswerOutcome(
+          explorationProgressController.submitAnswer(interactionObject)
+        )
+      }
+    } else {
+      moveToNextState()
+    }
+  }
+
+  override fun onPreviousButtonClicked() {
+    explorationProgressController.moveToPreviousState()
+  }
+
+  override fun onNextButtonClicked() {
+    moveToNextState()
+  }
+
+  private fun moveToNextState() {
+    checkAndUpdateOldStateNameList()
+    itemList.clear()
+    currentAnswerOutcome = null
+    explorationProgressController.moveToNextState()
+  }
+
+  private fun createContinueButtonAnswer(): InteractionObject {
+    return InteractionObject.newBuilder().setNormalizedString(DEFAULT_CONTINUE_INTERACTION_TEXT_ANSWER).build()
+  }
+
+  private fun checkAndUpdateOldStateNameList() {
+    if (currentAnswerOutcome != null
+      && !currentAnswerOutcome!!.sameState
+      && !oldStateNameList.contains(currentEphemeralState.get()!!.state.name)
```

Comment:
Why exactly do we need an old state name list? I'm a bit concerned about storing state--I actually tried to model the proto such that no additional state would be needed, so I'd like to understand the use case here (especially since any state stored in the UI may require reconstitution upon configuration change, where this isn't being retained currently).

Reply:
@BenHenning Follow up from comment on line 185

The oldstateList help us in identifying whether the `hasGeneralContinueButton` should be true or false to decide whether to show `Continue` button or `Next` button to move to next state from an already answer-submitted state.

Reply:
Ack. Per my latest reply on the line 185 comment thread, maybe we can also avoid this variable by fixing the missing information in the domain layer. Please make sure this specific case has a test so that I can verify my fix works when I clean this up later (assuming I understand the situation correctly).

Reply:
Added issue #257